src/dll/Hooks/ValidateScripts.cpp

```diff
@@ -89,3 +90,61 @@ bool Hooks::ValidateScripts::Detach()
     isAttached = result != NO_ERROR;
     return !isAttached;
 }
+
+std::wstring WritePopupMessage(const std::vector<ValidationError>& validationErrors,
+                               const std::vector<PluginSystem::PluginName>& incompatiblePlugins)
+{
+    if (validationErrors.empty())
+    {
+        return {};
+    }
+
+    std::wstring message;
+
+    if (!incompatiblePlugins.empty())
+    {
+        fmt::format_to(std::back_inserter(message),
+                       L"The following red4ext plugins could not be loaded because they are "
+                       L"incompatible with the current version of the game:\n");
+
+        for (const auto& plugin : incompatiblePlugins)
+        {
+            fmt::format_to(std::back_inserter(message), L"- {}\n", plugin);
+        }
+        fmt::format_to(std::back_inserter(message), L"\n");
+    }
+
+    std::unordered_set<std::string_view> faultyScriptFiles;
+
+    for (const auto& error : validationErrors)
+    {
+        const auto ref = error.GetSourceRef();
+        if (ref)
+        {
+            faultyScriptFiles.insert(ref->file);
+        }
+    }
+
+    if (!faultyScriptFiles.empty())
+    {
+        fmt::format_to(std::back_inserter(message),
+                       L"The following scripts contain invalid native definitions and will prevent "
+                       L"your game from starting:\n");
+
+        std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
+        for (const auto& file : faultyScriptFiles)
+        {
+            const auto fileWide = converter.from_bytes(std::string(file));
```

Comment:
this allocation is kinda pointless, but there seems to be no simple way to pass a `std::string_view` to the wstring converter

Reply:
oh... i didnt even see this :D

that is deprecated for various reasons and shouldnt be used... take a peek probably at UTF16ToUTF8 we use in CET.

it uses win32 api for this.

Reply:
writing a regular string seems to work fine though (without having to bother with the conversion):
```c++
  for (const auto& file : faultyScriptFiles)
  {
      fmt::format_to(std::back_inserter(message), "- {}\n", file);
  }
```
but I don't know if it handles multi byte characters correctly

Reply:
maybe even bopps has some helper utility function for this somewhere... is quite common to need such thing

Reply:
is a bit of format abuse i suppose but yeah, guess that should also work

Reply:
oh, he does, there's `std::wstring Widen(const std::string_view aText)`

Reply:
@WSSDude done

Reply:
wait a sec... i just realised - why is that a string_view? :suspect: 

isnt it a file name/path? is it a view to utf8string?

just making sure, dont exactly see from changes where it is filled and im on phone 😅

Reply:
@WSSDude it's a std::string_view around a UTF-8 encoded path, copied over from the Rust side